### PR TITLE
Check args dimensions for mesolve/sesolve

### PR DESCRIPTION
**Pull request by gautierronan**
A few basic checks of args dimensions inside mesolve/sesolve.

### CODE REVIEW

**Comment by pierreguilmin** [comment]
Nice! I'll review this weekend.

**Comment by pierreguilmin** [comment]
Various reformulations, see commit details. Note that this is a bit rudimentary because we just check parts of the shape and not the full specification (e.g. for `H` we should also check that it's of dim 2 or 3), we should check that `tsave` is ordered, etc... As discussed the improved version with `_checks.py` will adress all this. But this is good enough for now, and much better than having nothing 👌 @gautierronan I'll wait for your review before merging.

**Review by pierreguilmin — DISMISSED** [review]

File: dynamiqs/mesolve/mesolve.py
Comment (pierreguilmin) [reply dynamiqs/mesolve/mesolve.py:222]: We **do** accept both, it's just a bit clearer to have `list[ArrayLike]` than `ArrayLike`, so people see immediately they should not pass a single expectation operator. I agree it's a bit loose (proper typing would be `list[ArrayLike] | ArrayLike`, but `list[ArrayLike]` = `ArrayLike`, so proper typing would be `ArrayLike`), but I think it's clearer.

For now the doc says:
```
exp_ops (list of array-like, of shape (nE, n, n), optional) – List of operators for which the expectation value is computed.
```
For the exception, do you prefer
```
            'Argument `exp_ops` must be a list of operator with shape (nE, n, n), but has shape'
            f' exp_ops.shape={exp_ops.shape}.'
```

**Comment by gautierronan** [comment]
Converted to the new format with `check_shape` and `check_times`.

Also, I removed the necessity of time arrays to be only made of positive values, as I believe this is not true anymore.

**Review by pierreguilmin — APPROVED** [review]
Very cool! Thanks a lot for rewriting this with the new checks logic.

I've reworkd our `_astimearray()` function, currently the error is not satisfying if either `H` or `jump_ops` don't have the correct shape (because we were catching and interpreting the `ValueError`). Merging 🎉

**Pull request merged**
